cleanlab/object_detection/rank.py

```diff
@@ -0,0 +1,1049 @@
+
+"""Methods to rank and score images in an object detection dataset (object detection data), based on how likely they
+are to contain label errors. """
+
+import warnings
+from typing import List, Dict, Any, Optional, Tuple, TypedDict
+import numpy as np
+import copy
+
+AuxiliaryTypesDict = TypedDict(
+    "AuxiliaryTypesDict",
+    {
+        "pred_labels": np.ndarray,
+        "pred_label_probs": np.ndarray,
+        "pred_bboxes": np.ndarray,
+        "lab_labels": np.ndarray,
+        "lab_bboxes": np.ndarray,
+        "similarity_matrix": np.ndarray,
+        "min_possible_similarity": float,
+    },
+)
+
+
+def get_label_quality_scores(
+    labels: List[Dict[str, Any]],
+    predictions: List[np.ndarray],
+    *,
+    method: str = "subtype",
+    probability_threshold: Optional[float] = None,
+    verbose: bool = True,
+) -> np.ndarray:
+    """Computes a label quality score for each image in the dataset.
+
+    For object detection datasets, the label quality score for an image estimates how likely it has been correctly labeled.
+    Lower scores indicate images whose annotation is more likely imperfect.
+    Annotators may have mislabeled an image because they:
+    - overlooked an object (missing annotated bounding box),
+    - chose the wrong class label for an annotated box in the correct location,
+    - imperfectly annotated the location/edges of a bounding box.
+    Any of these annotation errors should lead to an image with a lower label quality score.
+
+    Score is between 0 and 1.
+
+        - 1 - clean label (given label is likely correct).
+        - 0 - dirty label (given label is likely incorrect).
+
+    A score is calculated for each of ``N`` images, with ``K`` total classes in the data.
+    Each image has ``L`` annotated bounding boxes and ``M`` predicted bounding boxes.
+
+    Parameters
+    ----------
+    labels:
+        A list of ``N`` dictionaries such that ``labels[i]`` contains the given labels for the `i`-th image in the format
+       ``{'bboxes': np.ndarray((M,4)), 'labels': np.ndarray((M,)), 'image_name': str}`` where ``L`` is the number of annotated bounding boxes
+       for the `i`-th image and ``bboxes[j]`` is in the format ``[x,y,x,y]`` with given label ``labels[j]``. (``image_name`` is optional here)
+
+    predictions:
+        A list of ``N`` ``np.ndarray`` such that ``predictions[i]`` corresponds to the model predictions for the `i`-th image
+        in the format ``np.ndarray((K,))`` and ``predictions[i][k]`` is of shape ``np.ndarray(M,5)``
+        where ``M`` is the number of predicted bounding boxes for class ``k`` and the five columns correspond to ``[x,y,x,y,pred_prob]`` returned
+        by the model.
+
+    method:
+        The method used to calculate label_quality_scores. Options:
+
+        - ``subtype_lqs``: calculates image score as a composite score of the quality of badly located, swapped and missing bounding boxes.
+
+    probability_threshold:
```

Comment:
Is this even still used at all in our final scoring method? I thought not but don't recall exactly.

Reply:
We never used it but later methods can maybe make use of it. Should we keep it in?

Reply:
no not if it's never used. We can always add more (keyword-only) args later, because of our usage of * to separate the keyword-only args from the positional args. Including args that have no effect should be avoided